### PR TITLE
[SPARK-38830][CORE] Warn on corrupted block messages

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to warn when `NettyBlockRpcServer` received a corrupted block RPC message or under attack.

- `IllegalArgumentException`: When the type is unknown/invalid when decoding. This fails at Spark layer.
- `NegativeArraySizeException`: When the size read is negative. This fails at Spark layer during buffer creation.
- `IndexOutOfBoundsException`: When the data field isn't matched with the size. This fails at Netty later.

### Why are the changes needed?

When the RPC messages are corrupted or the servers are under attack, Spark shows `IndexOutOfBoundsException` due to the failure from `Decoder`. Instead of `Exception`, we had better ignore the message with a directional warning message.
```
java.lang.IndexOutOfBoundsException:
    readerIndex(5) + length(602416) exceeds writerIndex(172):
UnpooledUnsafeDirectByteBuf(ridx: 5, widx: 172, cap: 172/172)
    at io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1477)
    at io.netty.buffer.AbstractByteBuf.checkReadableBytes(AbstractByteBuf.java:1463)
    at io.netty.buffer.UnpooledDirectByteBuf.readBytes(UnpooledDirectByteBuf.java:316)
    at io.netty.buffer.AbstractByteBuf.readBytes(AbstractByteBuf.java:904)
    at org.apache.spark.network.protocol.Encoders$Strings.decode(Encoders.java:45)
    at org.apache.spark.network.shuffle.protocol.UploadBlock.decode(UploadBlock.java:112)
    at org.apache.spark.network.shuffle.protocol.BlockTransferMessage$Decoder.fromByteBuffer(BlockTransferMessage.java:71)
    at org.apache.spark.network.netty.NettyBlockRpcServer.receive(NettyBlockRpcServer.scala:53)
    at org.apache.spark.network.server.TransportRequestHandler.processRpcRequest(TransportRequestHandler.java:161)
    at org.apache.spark.network.server.TransportRequestHandler.handle(TransportRequestHandler.java:109)
    at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:140)
    at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:53)

```

### Does this PR introduce _any_ user-facing change?

Yes, but this clarify the log messages from exceptions, `IndexOutOfBoundsException`.

### How was this patch tested?

Pass the CIs with newly added test suite.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/network/netty/NettyBlockRpcServerSuite.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+class NettyBlockRpcServerSuite extends SparkFunSuite {
+
+  test("SPARK-38830: Warn corrupted Netty messages") {
+    val serializer = new JavaSerializer(new SparkConf)
+    val server = new NettyBlockRpcServer("enhanced-rpc-server", serializer, null)
+    val bytes = Array[Byte](1.toByte)
+    val message = java.nio.ByteBuffer.wrap(bytes)
+    val client = mock(classOf[TransportClient])
+    server.receive(client, message)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/network/netty/NettyBlockRpcServerSuite.scala:34]: Previously, `IndexOutOfBoundsException` happens here.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala [hunk 2]
```diff
@@ -50,7 +51,17 @@ class NettyBlockRpcServer(
...
-    val message = BlockTransferMessage.Decoder.fromByteBuffer(rpcMessage)
+    val message = try {
+      BlockTransferMessage.Decoder.fromByteBuffer(rpcMessage)
+    } catch {
+      case _: IndexOutOfBoundsException =>
+        // Netty throws IndexOutOfBoundsException for corrupted buffers. In this case, we ignore
+        // the entire message with warnings because we cannot trust any contents.
+        logWarning(s"Ignored a corrupted RPC message (capacity: ${rpcMessage.capacity()}) " +
+          s"from ${client.getSocketAddress}. Please use `spark.authenticate.*` configurations " +
+          "in case of security incidents.")
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala:62]: If the incoming message is not valid, we can end up with `IllegalArgumentException`, `IndexOutOfBoundsException`, `IllegalArgumentException`, `NegativeArraySizeException`, etc.

Any particular reason to suspect `IndexOutOfBoundsException` ?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala:62]: Thank you for review. Yes, this code path has also more exception types like `RuntimeException` (due to `IOException`) and `UnsupportedOperationException`. We can add some of them more.

However, currently, this PR focused on random attacks which causes failures mostly at the first field, `appId` decoding which are observed during testing.
```
String appId = Encoders.Strings.decode(buf);
```

**Comment by dongjoon-hyun** [comment]
Hi, @mridulm . Do you want me to include more exceptions in this PR?

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala (hunk 2 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala:62]: Might be a good idea to expand to atleast include the common patterns we can observe. Top of my head, the following are possible:
* `IllegalArgumentException` - when the `type` is unknown/invalid when decoding.
* `NegativeArraySizeException` - if the size read is invalid (negative for example).
* `IndexOutOfBoundsException` - we are already handling it in this PR.

Rest can be enhanced in future works, thoughts ?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockRpcServer.scala:62]: Got i. Thank you for your advice, @mridulm .

**Review by mridulm — APPROVED** [review]
Thanks for fixing this @dongjoon-hyun !

**Comment by dongjoon-hyun** [comment]
Thank you so much, @mridulm !

**Comment by dongjoon-hyun** [comment]
Merged to master/3.3/3.2.

**Pull request closed without merging**